plugins/module_utils/identity/bitwarden.py

```diff
@@ -0,0 +1,321 @@
+        matches = [Item(dict_=d) for d in Client.run_json(query)
+                   if self.folder.matches(d) and d[attribute] == value]
+
+        if not matches:
+            raise KeyError('Item not found')
+        if len(matches) > 1:
+            raise AssertionError('Item not unique')
+
+        return matches[0]
+
+    def _dict_resolve_callback(self, value):
+        self._organization = Organization(id_=value['organizationId'])
+        self._folder = Folder(id_=value['folderId'])
+
+
+class Folder(_Element):
+    '''A folder in the password database.'''
+
+    organization = property(lambda self: self._organization)
+
+    def __init__(self, name=None, id_=None, organization=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify folder in multiple ways')
+
+        self._name = name
+        self._id = id_
+        self._organization = organization
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return folder dict where the specified attribute
+        has the specified value.
+        '''
+        matches = [folder for folder in self.organization.folders if folder.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Folder not found")
+        if len(matches) > 1:
+            raise AssertionError("Folder not unique")
+
+        return matches[0]
+
+    @property
+    def items(self):
+        query = ['list', 'items']
+
+        # Only restrict by folder if specified.
+        if self.name is not None:
+            # bw treats 'null' as items not belonging to any folder.
+            query.extend(['--folderid', self.id or 'null'])
+
+        # Filter by organization since `bw` ORs filters together.
+        return [Item(dict_=d) for d in Client.run_json(query) if self.organization.matches(d)]
+
+    def get_item(self, name=None, id_=None):
+        return Item(name, id_, self, self.organization)
+
+    def matches(self, folder):
+        '''Return whether folder matches a pattern indicated by self.
+
+        If self.name is None, any folder matches.
+        If self.name is '' or 'No Folder', only the null folder is matched.
+        All other folders are matched by id.
+        '''
+
+        if self.name is None:
+            return True
+
+        # No Folder is a special folder name in bw.
+        if self.name in ('', 'No Folder'):
+            return folder['folderId'] is None
+
+        return self.id == folder['folderId']
+
+
+class Organization(_Element):
+    '''An organization in the password database.'''
+
+    def __init__(self, name=None, id_=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify organization in multiple ways')
+
+        self._name = name
+        self._id = id_
+
+        if dict_:
+            self.dict = dict_
```

Comment:
Thanks! For clarification, you're saying `self.dict` isn't good practice, but `dict_` is okay?

I was aware one should avoid using built-ins' names (in Python or otherwise) is to avoid shadowing for new identifiers, which is why I've been using `id_` and `dict_`. In terms of python attributes, I didn't realize this was an issue since it never actually shadows anything since it never pollutes any namespaces directly, unlike, say, Java, where fields (I believe that is the correct equivalent to Python's attributes) are implicitly referenced, even without `this`.

Reply:
Well, there are (at least) two aspects to this. The first being avoid shadowing built-ins' names, yes, but the second one is semantics. Taking this example to an extreme (for the sake of argument), we should avoid calling variables `dict` or `dict_` for the same reason that we do not call other variables `int` or `int_`, or `str` or `float_`. Naming the variable after its *type* lacks *meaning* to it: 

_It is a `dict`, alright, but what is this `dict` carrying? Env variables? Attributes? A translation map?_ 

Hope you see the point: I reckon the good practice goes beyond naming conflicts.

Reply:
I agree; semantically, I'd like to provide as much detail as possible, but, to be honest, I have no idea what the dictionary really contains or can contain 😅 The documentation for the Bitwarden CLI only goes so far as to say it will be a JSON dictionary. I could call it `decoded_json` or something to that effect. Thoughts?

The documentation is so vague that, to create a new object in the vault, the documentation for the CLI states that one should call [`bw get template`](https://bitwarden.com/help/cli/#get-template)) and then fill it in using something like `jq`; there is no mention as to *how* to fill it in or what structure it Is guaranteed to take. It basically seems as if they're expecting a human to fill in the values (and I couldn't make head or tail of some of the features -- I'm not sure which fields are even supported in the free version of Bitwarden since there's a TOTP field, but that's reserved for premium accounts). I'm not sure how stable the Bitwarden CLI's API is, so I tried to minimize hard-coding semantic interpretations of the results.

Reply:
I am looking at the documentation, and it is a but odd indeed. I suggest you mark this module as WIP (just at that to the title as a prefix), and spend more time investigating the `bw` command and how it works.

One of the purposes of having an Ansible module is to make life easier for the user, to encapsulate complexity. I think this might be the case here, but I am also not familiar with BitWarden to tell.